extensions/druid-rocketmq/src/main/java/io/druid/firehose/rocketmq/RocketMQFirehoseFactory.java

```diff
@@ -0,0 +1,537 @@
+                Sets.newHashSet("feed")
+        );
+
+        final ByteBufferInputRowParser theParser = byteBufferInputRowParser.withParseSpec(
+                byteBufferInputRowParser.getParseSpec()
+                        .withDimensionsSpec(
+                                byteBufferInputRowParser.getParseSpec()
+                                        .getDimensionsSpec()
+                                        .withDimensionExclusions(
+                                                newDimExclus
+                                        )
+                        )
+        );
+
+        /**
+         * Topic-Queue mapping.
+         */
+        final ConcurrentHashMap<String, Set<MessageQueue>> topicQueueMap;
+
+        /**
+         * Default Pull-style client for RocketMQ.
+         */
+        final DefaultMQPullConsumer defaultMQPullConsumer;
+        final DruidPullMessageService pullMessageService;
+
+        messageQueueTreeSetMap.clear();
+        windows.clear();
+
+        try {
+            defaultMQPullConsumer = new DefaultMQPullConsumer(this.consumerGroup);
+            defaultMQPullConsumer.setMessageModel(MessageModel.CLUSTERING);
+            topicQueueMap = new ConcurrentHashMap<>();
+
+            pullMessageService = new DruidPullMessageService(defaultMQPullConsumer);
+            String[] topics = feed.split(",");
+            for (String topic : topics) {
+                topic = topic.trim();
+                if (topic.isEmpty()) {
+                    continue;
+                }
+                defaultMQPullConsumer.fetchSubscribeMessageQueues(topic);
+                topicQueueMap.put(topic, defaultMQPullConsumer.fetchMessageQueuesInBalance(topic));
+            }
+            DruidMessageQueueListener druidMessageQueueListener =
+                    new DruidMessageQueueListener(Sets.newHashSet(topics), topicQueueMap, defaultMQPullConsumer);
+            defaultMQPullConsumer.setMessageQueueListener(druidMessageQueueListener);
+            defaultMQPullConsumer.start();
+            pullMessageService.start();
+        } catch (MQClientException e) {
+            LOGGER.error("Failed to start DefaultMQPullConsumer", e);
+            throw new IOException("Failed to start RocketMQ client", e);
+        }
+
+        return new Firehose() {
+
+            @Override
+            public boolean hasMore() {
+                boolean hasMore = false;
+                DruidPullRequest earliestPullRequest = null;
+
+                for (Map.Entry<String, Set<MessageQueue>> entry : topicQueueMap.entrySet()) {
+                    for (MessageQueue messageQueue : entry.getValue()) {
+                        if (messageQueueTreeSetMap.keySet().contains(messageQueue)
+                                && !messageQueueTreeSetMap.get(messageQueue).isEmpty()) {
+                            hasMore = true;
+                        } else {
+                            DruidPullRequest newPullRequest = new DruidPullRequest();
```

Comment:
I think the answer is No. This request/notify/process/wait model is totally internal. And we have only one use case here.